x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/storage/ReactiveStorageDeciderService.java

```diff
@@ -255,7 +277,15 @@ && canAllocate(shard, allocation) == false) {
                 .mapToLong(this::sizeOf)
                 .sum();
 
-            return unallocatableBytes + unmovableBytes;
+            return new ShardsSize(
+                unallocatableBytes + unmovableBytes,
+                Stream.concat(
+                    unmovableShards.stream(),
+                    candidates.stream()
+                        .filter(Predicate.not(unmovableShards::contains))
+                        .filter(s -> cannotAllocateDueToStorage(s, allocation))
```

Comment:
This bit is somewhat expensive and we just did it above. I'd prefer to collect the unallocatable shards in a list to avoid the repeated calculation.